Initial support for AWS Batch Compute Environment

jonmer85 opened the pull request:
SUMMARY

This adds automation support for the following AWS Batch Compute Environments
ISSUE TYPE

Feature Pull Request
COMPONENT NAME

amazon module

ANSIBLE VERSION

ansible 2.2.2.0
config file =
configured module search path = Default w/o overrides

jonmeransonos committed:
Initial support for AWS Batch Compute Environment

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+
+    results = manage_state(module, aws)
+
+    module.exit_json(**results)
```
You can import camel_dict_to_snake_dict from module_utils/ec2 and use it here to maintain consistency with the rest of the AWS modules.

supreethSonos replied:
fixed.

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+            # Describe compute environment
+            response = get_current_compute_environment(module, aws)
+            if not response:
+                module.fail_json(msg='Unable to get compute environment information after creating')
```
Check mode fails for creation, though maybe that doesn't really matter?

supreethSonos replied:
added the condition while checking the check mode fails for creation.

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
```
I'm a little uncertain about supports_check_mode. It does correctly identify whether or not a change will be made and it does avoid actually making any changes, but the results don't accurately reflect what those changes will be. I'm not sure - other thoughts welcome.

supreethSonos replied:
Removed the supports_check_mode

s-hertel replied:
@supreethSonos I was wrong about this. Check mode just needs to return changed: true or false to accurately reflect that something will change. Diff mode is what I was thinking of. I have added back supports_check_mode=True.

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            state=dict(required=False, default='present', choices=['present', 'absent']),
```
Not necessary to specify required=False since it is the default.

supreethSonos replied:
fixed

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+    argument_spec.update(
+        dict(
+            state=dict(required=False, default='present', choices=['present', 'absent']),
+            compute_environment_name=dict(required=True, default=None),
```
Not necessary to specify default=None as that is the default.

supreethSonos replied:
fixed

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+    :param module_params:
+    :return:
+    """
+
```
Instead of making the snake_case to camel case functionality you can do something along the lines of
```
from ansible.module_utils.ec2 import snake_dict_to_camel_dict
...
...
api_params = snake_dict_to_camel_dict(dict(module.params))
fixed_params = {k:v for k,v in api_params.items() if v is not None}
return fixed_params
```
and remove cc()

supreethSonos replied:
fixed.

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+            if updates:
+                try:
+                    if not check_mode:
+                        response = aws.client().update_compute_environment(**compute_kwargs)
```
The response Metadata here doesn't seem very helpful. I think you could remove the 'response =' part of this line and then right below `response = get_current_compute_environment(module, aws)`

supreethSonos replied:
fixed.

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+
+  tags:
+    description:
+      - Key-value pair tags to be applied to resources that are launched in the compute environment.
```
Am I understanding right that tags can't be added or removed after creation?

jonmer85 replied:
Unfortunately that is correct. I believe you need to create a new Compute Environment to modify tags

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_together=[]
```
Can remove

supreethSonos replied:
fixed

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+
+RETURN = '''
+---
+batch_compute_environment_action:
```
currently returning changed, ansible_facts, and response

You can document the details like:
```
ansible_facts:
  description: helpful info
  type: complex
  returned: always
  contains:
    batch_compute_environment_action:
      description: describes what action was taken
    etc.
```

But I feel like what action was taken should be added to the returned response data (and not nest that data in response at all, since to access it the user will need to do something like registered_var.response.computeResources.tags[0] to get the associated tag) and ansible_facts should be removed. I don't know. Getting something called facts back from a module that isn't a _facts.py module doesn't seem intuitive. And the less nested the returned data is the easier to use.

ryansb replied:
Yeah I think most users would expect something like `registered_var.compute_resources.[attribute whatever]`

supreethSonos replied:
Removed ansible_facts. action taken is now returned in response data dictionary with the key batch_compute_environment_action and, is not nested in ansible_facts. See the sample in return doc string.

s-hertel commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+    """
+    Main entry point.
+
+    :return dict: ansible facts
```
returns more than that

supreethSonos replied:
fixed

jonmer85 commented:
@s-hertel Thank you for the feedback. Unfortunately I won't have the available time to make all of these changes and fully test them for a bit due to other commitments. Is there a release date for 2.4 that I should try to incorporate these changes before?

s-hertel commented:
No problem @jonmer85! The freeze date for modules is August 29th. The sooner you make the changes before that date the more likely it will make it into 2.4, since it will need to be re-reviewed (done on a best-effort basis).

ryansb commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+
+DOCUMENTATION = '''
+---
+module: batch_compute_environment
```
Can you please prefix this (and the others) with `aws_` since `batch_compute_environment` might be confused with some other batching system, and we have just one flat module namespace.

supreethSonos replied:
fixed

ryansb commented on lib/ansible/modules/cloud/amazon/batch_compute_environment.py:
```diff
@@ -0,0 +1,496 @@
+        environments = connection.client().describe_compute_environments(
+            computeEnvironments=[module.params['compute_environment_name']]
+        )
+        return environments['computeEnvironments'][0] if len(environments['computeEnvironments']) > 0 else None
```
Instead of one-line `if` can you use a standard `if/else` here?

supreethSonos replied:
fixed

supreethSonos committed:
Fixed comments.
changes include:
	1. modifying the doc string
	2. updating set_api_params method
	3. changing inline if/else to multi line
	4. modifying manage_state method
	5. removing default values in argument_spec in main method

supreethSonos committed:
Fixed comment
changes include:
	1. Modifying the final result using camel_dict_to_snake_dict to maintain consistency with the rest of the AWS modules. The keys in final result dict sometimes contain a mix of camel case and snake case in it. Converting and making sure all the keys are snake case here.

supreethSonos committed:
Fixed comments.
Modified the filename to aws_batch_compute_environemnt.py

supreethSonos committed:
Fixing ansibot errors:
changes include:
	1. modifying module name
	2. fixing the yml in RETURN
	3. changing the inline dict to be compatible with python 2.6

supreethSonos committed:
Fixed ansibot errors:
Removed blank spaces

supreethSonos committed:
Fixed ansibot errors:
changes include:
1. enclosing Python non-standard libraries in a try/except ImportError block.
2. fixing formatting errors

supreethSonos committed:
Fixing ansibot error
Removing ANSIBLE_METADATA as it’s not present in other modules such as rds_cluster, s3_event etc

supreethSonos committed:
Fixing ansibot error:
Adding ANSIBLE_METADATA with a newer version number

s-hertel committed:
Update copyright

Remove redundant exception handling around boto3_conn

Add traceback to exceptions

Use a new-style class

Remove HAS_BOTOCORE since HAS_BOTO3 handles it

Add ec2 documentation fragment

Region is not required because it may be found in environment vars

Document supports_check_mode

Add aliases for region

s-hertel committed:
Move common code for AWS batch modules in module_utils/aws/batch

ryansb commented on lib/ansible/modules/cloud/amazon/aws_batch_compute_environment.py:
```diff
@@ -0,0 +1,501 @@
+
+    results = manage_state(module, aws)
+
+    module.exit_json(**camel_dict_to_snake_dict(results))
```
This will also snakify tags, so someone with a tag `MyThing` would actually see it changed to `my_thing` so we should use the `ignore` option on camel2snake to avoid messing with user tags.

s-hertel committed:
Do not snake_case tags
